File: backend/scripts/utils/scrapeNewTaipeiParks.ts
import { parse } from 'csv-parse/sync';
import { existsSync, readFileSync, writeFileSync } from 'fs';
import { dirname, join } from 'path';
import { fileURLToPath } from 'url';
import { normalizeCityName, normalizeDistrictName, parseCityAndDistrict } from './parseAddress.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

export interface ParsedPlace {
  name: string;
  address: string;
  city: string | null;
  district: string | null;
  latitude: number | null;
  longitude: number | null;
  link?: string | null;
  metadata: Record<string, unknown>;
  source: string;
  sourceId: string;
}

/**
 * Scrape New Taipei City inclusive parks from HTML
 * HTML structure: <li id="view-..."> contains <h3 class="title"> and <p class="location">
 */
export function parseNewTaipeiParksHTML(filePath: string): ParsedPlace[] {
  if (!existsSync(filePath)) {
    console.warn(`File not found: ${filePath}`);
    return [];
  }

  const htmlContent = readFileSync(filePath, 'utf-8');
  return parseNewTaipeiParksHTMLFromContent(htmlContent, 0);
}

/**
 * Parse detailed park page HTML to extract address, description, coordinates, etc.
 */
export function parseParkDetailPage(htmlContent: string): {
  address: string;
  description: string;
  playEquipment: string; // 遊具設施
  fitnessEquipment: string; // 體健設施
  latitude: number | null;
  longitude: number | null;
  imageLinks: string[]; // 圖片連結
  metadata: Record<string, unknown>;
} {
  const baseUrl = 'https://www.ntparks.tw/';
  const result = {
    address: '',
    description: '',
    playEquipment: '',
    fitnessEquipment: '',
    latitude: null as number | null,
    longitude: null as number | null,
    imageLinks: [] as string[],
    metadata: {} as Record<string, unknown>,
  };

  // 提取位置（地址）
  const locationMatch = htmlContent.match(/<div[^>]*class="stitle subTitle_cray"[^>]*>位置<\/div>\s*<div[^>]*class="content"[^>]*>([^<]+)<\/div>/);
  if (locationMatch) {
    result.address = locationMatch[1].trim();
  }

  // 提取遊具設施
  const playEquipmentMatch = htmlContent.match(/<div[^>]*class="stitle subTitle_cray"[^>]*>遊具設施<\/div>\s*<div[^>]*class="content"[^>]*>([\s\S]*?)<\/div>/);
  if (playEquipmentMatch) {
    result.playEquipment = playEquipmentMatch[1].replace(/<[^>]+>/g, '').trim();
  }

  // 提取體健設施
  const fitnessEquipmentMatch = htmlContent.match(/<div[^>]*class="stitle subTitle_cray"[^>]*>體健設施<\/div>\s*<div[^>]*class="content"[^>]*>([\s\S]*?)<\/div>/);
  if (fitnessEquipmentMatch) {
    result.fitnessEquipment = fitnessEquipmentMatch[1].replace(/<[^>]+>/g, '').trim();
  }

  // 提取公園介紹
  const descriptionMatch = htmlContent.match(/<div[^>]*class="stitle subTitle_cray"[^>]*>公園介紹<\/div>\s*<div[^>]*class="content"[^>]*>([\s\S]*?)<\/div>/);
  if (descriptionMatch) {
    // 移除 HTML 標籤
    result.description = descriptionMatch[1].replace(/<[^>]+>/g, '').trim();
  }

  // 提取所有圖片連結（只提取公園相關的圖片，排除 logo、icon 等）
  // 格式: <img src="images/views/view-109-23-239/Photo-01.jpg">
  const imageMatches = htmlContent.matchAll(/<img[^>]*src="([^"]*images\/views\/[^"]+)"[^>]*>/g);
  for (const match of imageMatches) {
    let imageSrc = match[1].trim();
    // 移除開頭的空格（如果有）
    imageSrc = imageSrc.replace(/^\s+/, '');
    // 轉換成完整 URL
    if (imageSrc && !imageSrc.startsWith('http://') && !imageSrc.startsWith('https://')) {
      imageSrc = imageSrc.startsWith('/') ? `${baseUrl}${imageSrc.slice(1)}` : `${baseUrl}${imageSrc}`;
      result.imageLinks.push(imageSrc);
    }
  }

  // 從 Google Maps iframe 提取座標
  // 格式: !2d121.3315648886216!3d24.97092663816664 (經度, 緯度)
  const iframeMatch = htmlContent.match(/google\.com\/maps\/embed\?pb=[^"]*!2d([\d.]+)!3d([\d.]+)/);
  if (iframeMatch) {
    result.longitude = parseFloat(iframeMatch[1]);
    result.latitude = parseFloat(iframeMatch[2]);
  }

  // 提取其他資訊
  const infoMatches = htmlContent.matchAll(/<div[^>]*class="stitle subTitle_cray"[^>]*>([^<]+)<\/div>\s*<div[^>]*class="content"[^>]*>([\s\S]*?)<\/div>/g);
  for (const match of infoMatches) {
    const key = match[1].trim();
    const value = match[2].replace(/<[^>]+>/g, '').trim();
    if (key && value && key !== '位置' && key !== '公園介紹' && key !== '遊具設施' && key !== '體健設施') {
      result.metadata[key] = value;
    }
  }

  return result;
}

/**
 * Fetch HTML content from URL
 */
async function fetchHTML(url: string): Promise<string> {
  try {
    const response = await fetch(url);
    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }
    return await response.text();
  } catch (error) {
    console.error(`Failed to fetch ${url}:`, error);
    throw error;
  }
}

/**
 * Scrape all pages from the website (parallel processing)
 */
export async function scrapeAllPagesFromWeb(): Promise<ParsedPlace[]> {
  const baseUrl = 'https://www.ntparks.tw/';
  
  // 所有分頁：views.html, views02.html, ..., views16.html
  const pageUrls: string[] = [
    'views.html',
    'views02.html',
    'views03.html',
    'views04.html',
    'views05.html',
    'views06.html',
    'views07.html',
    'views08.html',
    'views09.html',
    'views10.html',
    'views11.html',
    'views12.html',
    'views13.html',
    'views14.html',
    'views15.html',
    'views16.html',
  ];

  console.log(`開始平行爬取 ${pageUrls.length} 個分頁...\n`);

  // 平行爬取所有分頁
  const pagePromises = pageUrls.map(async (pageFile, i) => {
    const pageUrl = `${baseUrl}${pageFile}`;
    console.log(`[${i + 1}/${pageUrls.length}] 開始爬取: ${pageUrl}`);
    
    try {
      const htmlContent = await fetchHTML(pageUrl);
      const places = parseNewTaipeiParksHTMLFromContent(htmlContent, i * 1000); // 使用偏移量避免 index 衝突
      console.log(`[${i + 1}/${pageUrls.length}] ✓ 找到 ${places.length} 個公園`);
      return places;
    } catch (error) {
      console.error(`[${i + 1}/${pageUrls.length}] ✗ 爬取失敗:`, error);
      return [] as ParsedPlace[];
    }
  });

  // 等待所有分頁爬取完成
  const allPagesPlaces = await Promise.all(pagePromises);
  const allPlaces = allPagesPlaces.flat();
  
  console.log(`\n所有分頁爬取完成，總共找到 ${allPlaces.length} 個公園`);
  console.log(`開始平行提取詳細資料...\n`);

  // 平行提取所有公園的詳細資料
  const detailPromises = allPlaces.map(async (place, index) => {
    if (!place.link) {
      return place;
    }

    try {
      const detailHtml = await fetchHTML(place.link);
      const detail = parseParkDetailPage(detailHtml);
      
      // 更新公園資料
      place.address = detail.address;
      place.latitude = detail.latitude;
      place.longitude = detail.longitude;
      
      // 合併 metadata（包含遊具設施、體健設施、公園介紹、圖片連結等）
      place.metadata = {
        ...place.metadata,
        description: detail.description,
        playEquipment: detail.playEquipment, // 遊具設施
        fitnessEquipment: detail.fitnessEquipment, // 體健設施
        imageLinks: detail.imageLinks, // 圖片連結陣列
        ...detail.metadata,
      };
      
      // 如果有地址，重新解析城市和區域
      if (detail.address) {
        const { city, district, remainingAddress } = parseCityAndDistrict(
          detail.address,
          '新北市',
          place.district
        );
        place.city = normalizeCityName(city);
        place.district = normalizeDistrictName(district);
        if (remainingAddress && !place.address.includes(remainingAddress)) {
          place.address = remainingAddress;
        }
      }
      
      if ((index + 1) % 50 === 0) {
        console.log(`  已處理 ${index + 1}/${allPlaces.length} 個公園...`);
      }
    } catch (error) {
      console.warn(`  無法提取 ${place.name} 的詳細資料:`, error);
    }
    
    return place;
  });

  // 等待所有詳細資料提取完成
  const placesWithDetails = await Promise.all(detailPromises);
  
  console.log(`\n✓ 所有詳細資料提取完成！`);
  return placesWithDetails;
}

/**
 * Parse HTML content (extracted from parseNewTaipeiParksHTML for reuse)
 */
function parseNewTaipeiParksHTMLFromContent(htmlContent: string, indexOffset: number = 0): ParsedPlace[] {
  const places: ParsedPlace[] = [];

  // 提取所有公園列表項
  const liMatches = htmlContent.matchAll(/<li[^>]*id="view-[^"]+"[^>]*>([\s\S]*?)<\/li>/g);
  
  let index = 0;
  for (const match of liMatches) {
    const content = match[1];
    
    // 提取公園名稱
    const titleMatch = content.match(/<h3[^>]*class="title"[^>]*>([^<]+)<\/h3>/);
    if (!titleMatch) continue;
    
    const name = titleMatch[1].trim();
    
    // 提取位置資訊（格式：新北市．區域）
    const locationMatch = content.match(/<p[^>]*class="location"[^>]*>([^<]+)<\/p>/);
    const location = locationMatch ? locationMatch[1].trim() : '';
    
    // 提取連結
    const linkMatch = content.match(/<a[^>]*href="([^"]+)"[^>]*class="views-List"/);
    let link = linkMatch ? linkMatch[1].trim() : null;
    
    // 將相對路徑轉換成完整 URL
    if (link && !link.startsWith('http://') && !link.startsWith('https://')) {
      const baseUrl = 'https://www.ntparks.tw/';
      link = link.startsWith('/') ? `${baseUrl}${link.slice(1)}` : `${baseUrl}${link}`;
    }
    
    // 解析位置資訊（格式：新北市．區域）
    let district: string | null = null;
    if (location.includes('．')) {
      const parts = location.split('．');
      if (parts.length >= 2) {
        district = parts[1].trim();
        if (district && !district.endsWith('區') && !district.endsWith('市') && !district.endsWith('鎮') && !district.endsWith('鄉')) {
          district = district + '區';
        }
      }
    }
    
    places.push({
      name,
      address: '', // 將從詳細頁面提取
      city: '新北市',
      district: normalizeDistrictName(district),
      latitude: null, // 將從詳細頁面提取
      longitude: null, // 將從詳細頁面提取
      link: link,
      metadata: {
        originalLocation: location,
        htmlId: match[0].match(/id="([^"]+)"/)?.[1] || null,
      },
      source: '新北市共融特色公園',
      sourceId: `ntpc_park_${indexOffset + index}_${name}`,
    });
    
    index++;
  }

  return places;
}

/**
 * Escape CSV field value
 */
function escapeCSVField(value: string | number | null | undefined): string {
  if (value === null || value === undefined) {
    return '';
  }
  const str = String(value);
  // If contains comma, quote, or newline, wrap in quotes and escape quotes
  if (str.includes(',') || str.includes('"') || str.includes('\n')) {
    return `"${str.replace(/"/g, '""')}"`;
  }
  return str;
}

/**
 * Convert ParsedPlace array to CSV format
 */
function convertToCSV(places: ParsedPlace[]): string {
  if (places.length === 0) {
    return '';
  }

  // CSV headers
  const headers = [
    'name',
    'address',
    'city',
    'district',
    'latitude',
    'longitude',
    'link',
    'playEquipment',
    'fitnessEquipment',
    'description',
    'imageLinks',
    'source',
    'sourceId',
  ];

  // Create CSV rows
  const rows = [headers.join(',')];

  for (const place of places) {
    // 從 metadata 中提取遊具設施、體健設施、公園介紹、圖片連結
    const playEquipment = (place.metadata.playEquipment as string) || '';
    const fitnessEquipment = (place.metadata.fitnessEquipment as string) || '';
    const description = (place.metadata.description as string) || '';
    const imageLinks = Array.isArray(place.metadata.imageLinks) 
      ? (place.metadata.imageLinks as string[]).join('; ') 
      : '';
    
    const row = [
      escapeCSVField(place.name),
      escapeCSVField(place.address),
      escapeCSVField(place.city),
      escapeCSVField(place.district),
      escapeCSVField(place.latitude),
      escapeCSVField(place.longitude),
      escapeCSVField(place.link),
      escapeCSVField(playEquipment),
      escapeCSVField(fitnessEquipment),
      escapeCSVField(description),
      escapeCSVField(imageLinks),
      escapeCSVField(place.source),
      escapeCSVField(place.sourceId),
    ];
    rows.push(row.join(','));
  }

  return rows.join('\n');
}

/**
 * Save parsed places to CSV file in data directory
 */
export function saveToCSV(
  places: ParsedPlace[], 
  filename: string = 'new-taipei-parks.csv',
  outputDir?: string
): void {
  const csvContent = convertToCSV(places);
  
  let outputPath: string;
  
  if (outputDir) {
    // 如果指定了輸出目錄，直接使用
    outputPath = join(outputDir, filename);
  } else {
    // 預設存到 data 目錄
    // Get the project root (assuming this file is in backend/scripts/utils)
    // Go up 3 levels: utils -> scripts -> backend -> root
    const projectRoot = join(__dirname, '../../..');
    const dataDir = join(projectRoot, 'data');
    outputPath = join(dataDir, filename);
  }

  writeFileSync(outputPath, csvContent, 'utf-8');
  console.log(`已將 ${places.length} 筆資料儲存至: ${outputPath}`);
}

/**
 * Parse HTML file and save all data to CSV in data directory
 */
export function parseAndSaveToCSV(
  htmlFilePath: string,
  csvFilename: string = 'new-taipei-parks.csv'
): void {
  const places = parseNewTaipeiParksHTML(htmlFilePath);
  if (places.length === 0) {
    console.warn('沒有找到任何公園資料');
    return;
  }
  
  // 存到 data 目錄
  const projectRoot = join(__dirname, '../../..');
  const dataDir = join(projectRoot, 'data');
  saveToCSV(places, csvFilename, dataDir);
}

/**
 * Scrape all pages from website and save to CSV
 */
export async function scrapeAllPagesAndSaveToCSV(
  csvFilename: string = 'new-taipei-parks.csv'
): Promise<void> {
  const places = await scrapeAllPagesFromWeb();
  if (places.length === 0) {
    console.warn('沒有找到任何公園資料');
    return;
  }
  
  // 存到 data 目錄
  const projectRoot = join(__dirname, '../../..');
  const dataDir = join(projectRoot, 'data');
  saveToCSV(places, csvFilename, dataDir);
}

/**
 * Parse New Taipei Parks CSV file
 */
interface NewTaipeiParksCSVRow {
  name: string;
  address: string;
  city: string;
  district: string;
  latitude: string;
  longitude: string;
  link?: string;
  playEquipment?: string;
  fitnessEquipment?: string;
  description?: string;
  imageLinks?: string;
  source: string;
  sourceId: string;
}

export function parseNewTaipeiParksCSV(filePath: string): ParsedPlace[] {
  if (!existsSync(filePath)) {
    console.warn(`File not found: ${filePath}`);
    return [];
  }

  const fileContent = readFileSync(filePath, 'utf-8');

  const records: NewTaipeiParksCSVRow[] = parse(fileContent, {
    columns: true,
    skip_empty_lines: true,
    bom: true,
  });

  const places: ParsedPlace[] = [];
  
  for (const row of records) {
    const lat = parseFloat(row.latitude);
    const lng = parseFloat(row.longitude);

    // Skip invalid coordinates
    if (isNaN(lat) || isNaN(lng) || lat < 20 || lat > 26 || lng < 118 || lng > 123) {
      continue;
    }

    // 解析圖片連結（用分號分隔）
    const imageLinks = row.imageLinks 
      ? row.imageLinks.split(';').map(link => link.trim()).filter(link => link.length > 0)
      : [];

    const place: ParsedPlace = {
      name: row.name || '未命名公園',
      address: row.address || '',
      city: normalizeCityName(row.city),
      district: normalizeDistrictName(row.district),
      latitude: lat,
      longitude: lng,
      link: row.link || null,
      metadata: {
        playEquipment: row.playEquipment || '',
        fitnessEquipment: row.fitnessEquipment || '',
        description: row.description || '',
        imageLinks: imageLinks,
      },
      source: row.source || '新北市共融特色公園',
      sourceId: row.sourceId || `ntpc_park_${row.name}_${lat}_${lng}`,
    };
    
    places.push(place);
  }
  
  return places;
}

File: data/scripts/scrape_new_taipei_parks.py
"""
爬取新北市共融特色公園資料
"""

import csv
import re
import os
from pathlib import Path
from typing import List, Dict, Any, Optional
import requests
from bs4 import BeautifulSoup
from .parse_address import normalize_city_name, normalize_district_name, parse_city_and_district


class ParsedPlace:
    """解析後的地點資料"""
    def __init__(
        self,
        name: str,
        address: str,
        city: Optional[str],
        district: Optional[str],
        latitude: Optional[float],
        longitude: Optional[float],
        link: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None,
        source: str = '',
        source_id: str = ''
    ):
        self.name = name
        self.address = address
        self.city = city
        self.district = district
        self.latitude = latitude
        self.longitude = longitude
        self.link = link
        self.metadata = metadata or {}
        self.source = source
        self.source_id = source_id

    def to_dict(self) -> Dict[str, Any]:
        """轉換為字典格式"""
        return {
            'name': self.name,
            'address': self.address,
            'city': self.city,
            'district': self.district,
            'latitude': self.latitude,
            'longitude': self.longitude,
            'link': self.link,
            'metadata': self.metadata,
            'source': self.source,
            'sourceId': self.source_id,
        }


def parse_park_detail_page(html_content: str) -> Dict[str, Any]:
    """
    解析詳細公園頁面 HTML 以提取地址、描述、座標等
    
    Returns:
        包含地址、描述、遊具設施、體健設施、座標、圖片連結等的字典
    """
    base_url = 'https://www.ntparks.tw/'
    result = {
        'address': '',
        'description': '',
        'playEquipment': '',  # 遊具設施
        'fitnessEquipment': '',  # 體健設施
        'latitude': None,
        'longitude': None,
        'imageLinks': [],  # 圖片連結
        'metadata': {},
    }

    soup = BeautifulSoup(html_content, 'html.parser')

    # 提取位置（地址）
    location_div = soup.find('div', class_='stitle subTitle_cray', string='位置')
    if location_div:
        content_div = location_div.find_next_sibling('div', class_='content')
        if content_div:
            result['address'] = content_div.get_text(strip=True)

    # 提取遊具設施
    play_equipment_div = soup.find('div', class_='stitle subTitle_cray', string='遊具設施')
    if play_equipment_div:
        content_div = play_equipment_div.find_next_sibling('div', class_='content')
        if content_div:
            result['playEquipment'] = content_div.get_text(strip=True)

    # 提取體健設施
    fitness_equipment_div = soup.find('div', class_='stitle subTitle_cray', string='體健設施')
    if fitness_equipment_div:
        content_div = fitness_equipment_div.find_next_sibling('div', class_='content')
        if content_div:
            result['fitnessEquipment'] = content_div.get_text(strip=True)

    # 提取公園介紹
    description_div = soup.find('div', class_='stitle subTitle_cray', string='公園介紹')
    if description_div:
        content_div = description_div.find_next_sibling('div', class_='content')
        if content_div:
            result['description'] = content_div.get_text(strip=True)

    # 提取所有圖片連結（只提取公園相關的圖片，排除 logo、icon 等）
    for img in soup.find_all('img', src=re.compile(r'images/views/')):
        image_src = img.get('src', '').strip()
        if image_src and not image_src.startswith(('http://', 'https://')):
            image_src = image_src.lstrip('/')
            image_src = f"{base_url}{image_src}" if not image_src.startswith(base_url) else image_src
            result['imageLinks'].append(image_src)

    # 從 Google Maps iframe 提取座標
    # 格式: !2d121.3315648886216!3d24.97092663816664 (經度, 緯度)
    iframe_match = re.search(r'google\.com/maps/embed\?pb=[^"]*!2d([\d.]+)!3d([\d.]+)', html_content)
    if iframe_match:
        result['longitude'] = float(iframe_match.group(1))
        result['latitude'] = float(iframe_match.group(2))

    # 提取其他資訊
    for div in soup.find_all('div', class_='stitle subTitle_cray'):
        key = div.get_text(strip=True)
        if key in ('位置', '公園介紹', '遊具設施', '體健設施'):
            continue
        
        content_div = div.find_next_sibling('div', class_='content')
        if content_div:
            value = content_div.get_text(strip=True)
            if key and value:
                result['metadata'][key] = value

    return result


def fetch_html(url: str) -> str:
    """從 URL 獲取 HTML 內容"""
    try:
        response = requests.get(url, timeout=30)
        response.raise_for_status()
        return response.text
    except Exception as e:
        print(f'Failed to fetch {url}: {e}')
        raise


def parse_new_taipei_parks_html_from_content(html_content: str, index_offset: int = 0) -> List[ParsedPlace]:
    """
    從 HTML 內容解析新北市公園列表
    
    Args:
        html_content: HTML 內容
        index_offset: 索引偏移量（用於多頁爬取時避免 index 衝突）
    
    Returns:
        解析後的地點列表
    """
    places = []
    soup = BeautifulSoup(html_content, 'html.parser')

    # 提取所有公園列表項
    li_items = soup.find_all('li', id=re.compile(r'^view-'))

    for index, li in enumerate(li_items):
        # 提取公園名稱
        title_tag = li.find('h3', class_='title')
        if not title_tag:
            continue

        name = title_tag.get_text(strip=True)

        # 提取位置資訊（格式：新北市．區域）
        location_tag = li.find('p', class_='location')
        location = location_tag.get_text(strip=True) if location_tag else ''

        # 提取連結
        link_tag = li.find('a', class_='views-List')
        link = link_tag.get('href') if link_tag else None

        # 將相對路徑轉換成完整 URL
        if link and not link.startswith(('http://', 'https://')):
            base_url = 'https://www.ntparks.tw/'
            link = link.lstrip('/')
            link = f"{base_url}{link}" if not link.startswith(base_url) else link

        # 解析位置資訊（格式：新北市．區域）
        district = None
        if '．' in location:
            parts = location.split('．')
            if len(parts) >= 2:
                district = parts[1].strip()
                if district and not district.endswith(('區', '市', '鎮', '鄉')):
                    district = district + '區'

        # 提取 HTML ID
        html_id = li.get('id', '')

        place = ParsedPlace(
            name=name,
            address='',  # 將從詳細頁面提取
            city='新北市',
            district=normalize_district_name(district),
            latitude=None,  # 將從詳細頁面提取
            longitude=None,  # 將從詳細頁面提取
            link=link,
            metadata={
                'originalLocation': location,
                'htmlId': html_id,
            },
            source='新北市共融特色公園',
            source_id=f'ntpc_park_{index_offset + index}_{name}',
        )

        places.append(place)

    return places


def parse_new_taipei_parks_html(file_path: str) -> List[ParsedPlace]:
    """
    從 HTML 檔案解析新北市公園資料
    
    Args:
        file_path: HTML 檔案路徑
    
    Returns:
        解析後的地點列表
    """
    if not os.path.exists(file_path):
        print(f'警告: File not found: {file_path}')
        return []

    with open(file_path, 'r', encoding='utf-8') as f:
        html_content = f.read()

    return parse_new_taipei_parks_html_from_content(html_content, 0)


def parse_new_taipei_parks_csv(file_path: str) -> List[ParsedPlace]:
    """
    解析新北市公園 CSV 檔案
    
    Args:
        file_path: CSV 檔案路徑
    
    Returns:
        解析後的地點列表
    """
    if not os.path.exists(file_path):
        print(f'警告: File not found: {file_path}')
        return []

    places = []

    with open(file_path, 'r', encoding='utf-8-sig') as f:
        reader = csv.DictReader(f)
        for row in reader:
            try:
                lat = float(row.get('latitude', 0))
                lng = float(row.get('longitude', 0))
            except (ValueError, TypeError):
                continue

            # 跳過無效座標（台灣地區：緯度約 21-26，經度約 118-123）
            if lat < 20 or lat > 26 or lng < 118 or lng > 123:
                continue

            # 解析圖片連結（用分號分隔）
            image_links = []
            if row.get('imageLinks'):
                image_links = [link.strip() for link in row['imageLinks'].split(';') if link.strip()]

            place = ParsedPlace(
                name=row.get('name', '未命名公園'),
                address=row.get('address', ''),
                city=normalize_city_name(row.get('city')),
                district=normalize_district_name(row.get('district')),
                latitude=lat,
                longitude=lng,
                link=row.get('link'),
                metadata={
                    'playEquipment': row.get('playEquipment', ''),
                    'fitnessEquipment': row.get('fitnessEquipment', ''),
                    'description': row.get('description', ''),
                    'imageLinks': image_links,
                },
                source=row.get('source', '新北市共融特色公園'),
                source_id=row.get('sourceId', f"ntpc_park_{row.get('name', '')}_{lat}_{lng}"),
            )

            places.append(place)

    return places


def escape_csv_field(value: Any) -> str:
    """轉義 CSV 欄位值"""
    if value is None:
        return ''
    str_value = str(value)
    # 如果包含逗號、引號或換行，用引號包裹並轉義引號
    if ',' in str_value or '"' in str_value or '\n' in str_value:
        return f'"{str_value.replace('"', '""')}"'
    return str_value


def convert_to_csv(places: List[ParsedPlace]) -> str:
    """將地點列表轉換為 CSV 格式"""
    if not places:
        return ''

    # CSV headers
    headers = [
        'name', 'address', 'city', 'district', 'latitude', 'longitude',
        'link', 'playEquipment', 'fitnessEquipment', 'description',
        'imageLinks', 'source', 'sourceId'
    ]

    rows = [','.join(headers)]

    for place in places:
        # 從 metadata 中提取遊具設施、體健設施、公園介紹、圖片連結
        play_equipment = place.metadata.get('playEquipment', '')
        fitness_equipment = place.metadata.get('fitnessEquipment', '')
        description = place.metadata.get('description', '')
        image_links = place.metadata.get('imageLinks', [])
        image_links_str = '; '.join(image_links) if isinstance(image_links, list) else ''

        row = [
            escape_csv_field(place.name),
            escape_csv_field(place.address),
            escape_csv_field(place.city),
            escape_csv_field(place.district),
            escape_csv_field(place.latitude),
            escape_csv_field(place.longitude),
            escape_csv_field(place.link),
            escape_csv_field(play_equipment),
            escape_csv_field(fitness_equipment),
            escape_csv_field(description),
            escape_csv_field(image_links_str),
            escape_csv_field(place.source),
            escape_csv_field(place.source_id),
        ]
        rows.append(','.join(row))

    return '\n'.join(rows)


def save_to_csv(places: List[ParsedPlace], filename: str = 'new-taipei-parks.csv', output_dir: Optional[str] = None) -> None:
    """將地點列表儲存為 CSV 檔案"""
    csv_content = convert_to_csv(places)

    if output_dir:
        output_path = Path(output_dir) / filename
    else:
        # 預設存到 data 目錄
        # 假設此檔案在 data/scripts/，向上兩層到專案根目錄
        project_root = Path(__file__).parent.parent.parent
        data_dir = project_root / 'data'
        output_path = data_dir / filename

    output_path.parent.mkdir(parents=True, exist_ok=True)
    with open(output_path, 'w', encoding='utf-8') as f:
        f.write(csv_content)

    print(f'已將 {len(places)} 筆資料儲存至: {output_path}')
